### PR TITLE
feat: make cache server stateless & native multipart uploads

### PR DESCRIPTION
**Pull request by LouisHaftmann**

### CODE REVIEW

File: lib/db/migrations.ts [hunk 1]
```diff
@@ -24,5 +24,38 @@ export function migrations(dbType: DatabaseDriverName) {
         await db.schema.dropTable('cache_keys').ifExists().execute()
       },
     },
+    '2024-10-15T16:39:29': {
```
Comment (DrJume) [inline lib/db/migrations.ts:25]: maybe give it a name, so that we know in the future why it was needed

File: lib/storage/drivers/s3.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+import { defineStorageDriver } from '~/lib/storage/defineStorageDriver'
+import { streamToBuffer } from '~/lib/utils'
+
+export const s3Driver = defineStorageDriver({
+  envSchema: z.object({
+    STORAGE_S3_BUCKET: z.string().min(1),
+    STORAGE_S3_ENDPOINT: z.string().min(1),
+    STORAGE_S3_REGION: z.string().min(1).default(Minio.DEFAULT_REGION),
+    STORAGE_S3_PORT: z.coerce.number().positive(),
+    STORAGE_S3_USE_SSL: z.string().transform((v) => v === 'true'),
```
Comment (DrJume) [inline lib/storage/drivers/s3.ts:24]: use booleanSchema

File: lib/env.ts [hunk 3]
```diff
@@ -1,6 +1,10 @@
 import { z } from 'zod'
 
+const booleanSchema = z.string().transform((v) => v.toLowerCase() === 'true')
+
 const envSchema = z.object({
+  ENABLE_DIRECT_DOWNLOADS: booleanSchema.default('false'),
```
Comment (DrJume) [inline lib/env.ts:6]: is this used somewhere?

File: lib/storage/drivers/filesystem.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            force: true,
+          })
+        }
+      },
+    }
+  },
+})
+
+function getUploadBufferPath(uploadId: string) {
+  return path.join(ENV.TEMP_DIR, 'github-actions-cache-server', uploadId)
```
Comment (DrJume) [inline lib/storage/drivers/filesystem.ts:74]: there is a fs api for creating tmp dirs. Maybe would better support other unix compatible systems (or windows)
https://nodejs.org/api/fs.html#fsmkdtempprefix-options-callback

File: lib/storage/index.ts [hunk 5]
```diff
@@ -84,7 +84,7 @@ export async function initializeStorage() {
           .values({
             created_at: new Date().toISOString(),
             driver_upload_id: driverUploadId,
-            id: uploadId,
+            id: uploadId.toString(),
```
Comment (DrJume) [inline lib/storage/index.ts:87]: instead .toString() above?

Comment (LouisHaftmann) [reply lib/storage/index.ts:87]: the `uploadId` returned by `reserveCache` needs to be a number

**Comment by LouisHaftmann** [comment]
will update docs before merging

File: lib/storage/index.ts (hunk 5 above)
Comment (DrJume) [reply lib/storage/index.ts:87]: Why? Just asking

Comment (LouisHaftmann) [reply lib/storage/index.ts:87]: github's package expects a number

**Comment by LouisHaftmann** [comment]
will update docs in separate branch

**Pull request merged**